oaipmh/processors/db.py

```diff
@@ -1,23 +1,147 @@
+        .subquery()
+    )
+
+    metadata_alias = aliased(Metadata, data_subquery)
+    if all_versions:
+        result = Session.query(metadata_alias).all()
+    else:
+        result = (
+            Session.query(metadata_alias)
+            .filter(metadata_alias.is_current == 1)
+            .first()
+        )
+    return result
 
+def get_list_data(just_ids:bool, start_date :datetime, end_date:datetime, all_versions: bool, rq_set:Optional[Union[Group, Archive, Category]], skip:int, limit: int)->List[Metadata]:
+    """fetches list of data according to given parameters"""
+    #updated is sometimes NULL so we are using modtime
+    start_timestamp=start_date.timestamp()
+    end_timestamp=end_date.timestamp()
 
-def get_record_data_current(arxiv_id: Identifier )-> Optional[Metadata]:
-    """fetch latest metadata for a specific paper"""
-    data=(Session.query(Metadata)
-          .filter(Metadata.paper_id == arxiv_id.id)
-          .filter(Metadata.is_current==1)
-          .first()
+    #all papers that have been updated within the time frame
+    doc_ids=( 
+        Session.query(Metadata.document_id)
+        .filter(
+            Metadata.modtime >= start_timestamp,  
+            Metadata.modtime <= end_timestamp,  
+            Metadata.is_current == 1,  
+        )
+        .subquery()
     )
-    return data
+
+    #filter for certain categories
+    if rq_set:
+        archives, cats=process_requested_subject(rq_set)
+        aic = aliased(t_arXiv_in_category)
+        cat_conditions = [and_(aic.c.archive == arch_part, aic.c.subject_class == subj_part) for arch_part, subj_part in cats]
+        doc_ids=(Session.query(doc_ids.c.document_id)
+            .join(aic, doc_ids.c.document_id == aic.c.document_id)
+            .filter(
+                or_(
+                    aic.c.archive.in_(archives),
+                    or_(*cat_conditions),
+                )
+            )
+            .subquery()
+        )   
     
-def get_record_data_all(arxiv_id: Identifier)-> Optional[List[Metadata]]:
-    """fetch all metadata for a specific paper"""
-    data=(Session.query(Metadata)
-          .filter(Metadata.paper_id == arxiv_id.id)
-          .all()
+    #select exact group of documents that will be reported on
+    selected_doc_ids=(
+        Session.query(Metadata.document_id)
+        .filter(Metadata.document_id.in_(doc_ids.select()))
+        .filter(Metadata.is_current == 1)
+        .order_by(Metadata.modtime, Metadata.paper_id)
+        .offset(skip)
+        .limit(limit+1) #one extra to see if resumption token needed
+        .subquery()
     )
-    return data
\ No newline at end of file
+
+    #fetch the metadata
+    if just_ids: #only need the data for the header portion
+        data = (
+            Session.query(Metadata)
+            .filter(Metadata.document_id.in_(selected_doc_ids.select()), 
+                    Metadata.is_current == 1
+            )
+            .options(load_only(
```

Comment:
FYI I think that `.options` can be called later so you can have the same query but with different `load_only` options.